server/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -434,7 +433,7 @@ public synchronized IndexShard createShard(
         ShardLock lock = null;
         eventListener.beforeIndexShardCreated(routing, indexSettings);
         try {
-            lock = nodeEnv.shardLock(shardId, "starting shard", TimeUnit.SECONDS.toMillis(5));
+            lock = nodeEnv.shardLock(shardId, "starting shard", 0L);
```

Comment:
We can use `shardLock(ShardId id, final String details)` here. I think a comment describing that the shard creation is retried would be useful too